### ISSUE TITLE
Error during reference building

### ISSUE THREAD

**Issue by mrashwing** [williamritchie/IRFinder#93]
I'm getting the following error while building the hg19 reference with IRFinder 1.3.0 (the build works great on IRFinder 1.2.6).
Output of job.err file (I'm running this on a cluster):
`[bam_sort_core] merging from 60 files and 30 in-memory blocks...
gzip: tmp_14769/HSCHR21_1_CTG1_1.bed.lzop: not in gzip format
gzip: tmp_14769/GL000204.1.bed.lzop: not in gzip format
gzip: tmp_14769/GL000202.1.bed.lzop: not in gzip format
gzip: tmp_14769/HSCHR15_1_CTG8.bed.lzop: not in gzip format
gzip: tmp_14769/HG990_PATCH.bed.lzop: not in gzip format
gzip: tmp_14769/HSCHR17_3_CTG4.bed.lzop: not in gzip format
gzip: tmp_14769/HG280_PATCH.bed.lzop: not in gzip format
...(many more like this)`
The contents of job.out are as follows:
`Launching reference build process. The full build might take hours.
<Phase 1: STAR Reference Preparation>
May 21 00:05:11 ..... started STAR run
May 21 00:05:11 ... starting to generate Genome files
May 21 00:13:39 ... starting to sort Suffix Array. This may take a long time...
May 21 00:15:10 ... sorting Suffix Array chunks and saving them to disk...
May 21 02:25:48 ... loading chunks from disk, packing SA...
May 21 02:27:44 ... finished generating suffix array
May 21 02:27:44 ... generating Suffix Array index
May 21 02:31:51 ... completed Suffix Array index
May 21 02:31:51 ..... processing annotations GTF
May 21 02:32:09 ..... inserting junctions into the genome indices
May 21 02:36:53 ... writing Genome to disk ...
May 21 02:37:06 ... writing Suffix Array to disk ...
May 21 02:37:19 ... writing SAindex to disk
May 21 02:37:22 ..... finished successfully
<Phase 2: Mapability Calculation>
May 21 02:37:25 ... mapping genome fragments back to genome...
May 21 06:31:07 ... sorting aligned genome fragments...
May 21 06:37:25 ... indexing aligned genome fragments...
May 21 06:38:38 ... filtering aligned genome fragments by chromosome/scaffold...
May 21 06:42:22 ... merging filtered genome fragments...
May 21 06:42:25 ... calculating regions for exclusion...
May 21 06:44:51 ... cleaning temporary files...
<Phase 3: IRFinder Reference Preparation>
May 21 06:44:52 ... building Ref 1...
May 21 06:45:30 ... building Ref 2...
May 21 06:45:30 ... building Ref 3...
May 21 06:45:30 ... building Ref 4...
May 21 06:45:45 ... building Ref 5...
May 21 06:46:05 ... building Ref 6...
May 21 06:46:05 ... building Ref 7...
May 21 06:46:05 ... building Ref 8...
May 21 06:46:06 ... building Ref 9...
May 21 06:46:06 ... building Ref 10b...
May 21 06:46:06 ... building Ref 11b...
May 21 06:46:06 ... building Ref 12b...
Error: ref-cover.bed is empty.
Error: ref-read-continues.ref is empty.
Error: IRFinder reference building FAILED.`
On checking the Log.out from logSTARbuild folder, the alignment seems to have completed successfully. There is no log file in the Mappability folder.  Using star/2.7.2b bedtools/2.29.2 samtools/1.6. Any help appreciated!

**Comment by dg520** [comment]
Hi @mrashwing ,

This is weird. Do you have `/usr/bin/lzop` executable? IRFinder defaults in using `gzip` and creating those temporary bed files in `.gz` format. It's only when `/usr/bin/lzop` executable exists, IRFinder will use `lzop` and `.lzop`. This particular part has the same behavior and code as 1.2.6. Nothing changed. 

In your case, IRFinder still used `gzip`, meaning  `/usr/bin/lzop` was not found. But it somehow used `.lzop` as extension, which should only happen when `/usr/bin/lzop` was there. I'm not sure how that occurred on your end. I cannot reproduce it on my end.

The bottom line is, the IRFinder always uses the same reference since the first release. You can use the reference prepared by 1.2.6. That might be the quickest fix for you. Alternatively, we can comment out Line 19-23 in `bin/util/Mapability`  and re-run 1.3.0 reference preparation again (you can use `BuildRefFromSTARRef` mode to skip rebuilding the STAR reference) to see if it works.

If you choose the latter step, could you please let me know your feedback. Appreciate that. 

Best,
Dadi

**Comment by dg520** [comment]
Hi @mrashwing ,

Found the problem. In the new Mapability algorithm part, the merging step is hard-coded in `zcat`, which only works for `.gz` files. I believe that's where the error comes in. 
If you commented out Line 19-23 in `bin/util/Mapability`, it will force IRFinder only using `gzip`. Things should work again. 

I don't have `lzop` installed on my machine. I'm sorry that's not tested widely and I'll give it a fix in the coming days. Please bear with the temporary fixation and thanks for reporting this bug.

Best,
Dadi

**Comment by mrashwing** [comment]
Great! I'll try it out and update. BTW, thanks for the _mergesort fix in version 1.3.0, I've been manually editing the line to use stable (I have been wondering why no one encountered the mergesort compatibility problems so far)

**Comment by dg520** [comment]
Hi @mrashwing ,

I just released an updated version. It should now work with the `lzop` path as well. Would you mind try and let me know?

For that Perl problem, I guess a lot of us are using some old versions of Linux Distro where Perl is far behind the latest version. I occasionally encountered that problem on my home computer after an LTS update.

Best,
Dadi

**Comment by mrashwing** [comment]
Update: 
I think the build is working now, both with the new version of IRFinder and by commenting out lines 19-23 from Mapability (I used `BuildRefFromSTARRef`). However the log files are sligtly different from 1.2.6. In v1.2.6, a successful reference build used to give me the following:
`$ cat job_1719932.out
Launching reference build process. The full build should take at least one hour.
Usage : /project/SCCC/Conrad_lab/s194063/software/irfinder-1.2.6/bin/util/IRFinder-BuildRefFromEnsembl mode threads STAR-executable base_ftp_url_of_ensembl_genome+gtf output_directory(must not exist) additional_genome_reference(eg: ERCC) non_polyA_genes-as-bed region_blacklist-as-bed
Usage example: /project/SCCC/Conrad_lab/s194063/software/irfinder-1.2.6/bin/util/IRFinder-BuildRefFromEnsembl BuildRef 12 STAR "ftp://ftp.ensembl.org/pub/release-75/fasta/homo_sapiens/dna/" "IRFinder/REF/Human" "Refernce-ERCC.fa.gz" [non_polyA_genes.bed] [blacklist.bed]
May 21 07:38:03 ..... started STAR run
May 21 07:38:03 ... starting to generate Genome files
May 21 07:45:29 ... starting to sort Suffix Array. This may take a long time...
May 21 07:46:50 ... sorting Suffix Array chunks and saving them to disk...
May 21 09:16:33 ... loading chunks from disk, packing SA...
May 21 09:18:46 ... finished generating suffix array
May 21 09:18:46 ... generating Suffix Array index
May 21 09:22:35 ... completed Suffix Array index
May 21 09:22:35 ..... processing annotations GTF
May 21 09:22:49 ..... inserting junctions into the genome indices
May 21 09:26:05 ... writing Genome to disk ...
May 21 09:26:46 ... writing Suffix Array to disk ...
May 21 09:27:23 ... writing SAindex to disk
May 21 09:27:25 ..... finished successfully
Star genome build result: 0
Commence STAR mapping run for mapability.
Thu May 21 09:27:28 CDT 2020
Completed STAR run.
Thu May 21 12:06:41 CDT 2020
Commence Coverage calculation.
Completed coverage exclusion calculation.
Thu May 21 13:18:44 CDT 2020
Mapability result: 0
Build Ref 1
Build Ref 2
Build Ref 3
Build Ref 4
Build Ref 5
Build Ref 6
Build Ref 7
Build Ref 8
Build Ref 9
Build Ref 10b
Build Ref 11b
Build Ref 12b
**COMPLETE
Ref build result: 0
ALL DONE**`
and the following for the job.err:
`$ cat job_1719932.err

real    159m12.398s
user    4689m6.437s
sys     55m16.162s

real    71m59.675s
user    2588m45.517s
sys     802m26.090s

real    0m3.268s
user    0m4.819s
sys     0m0.228s`
In v1.3.0, I get the following:
`$cat job_1722418.out
Launching reference build process. The full build might take hours.
<Phase 1: STAR Reference Preparation>
May 22 14:59:38 ..... started STAR run
May 22 14:59:38 ... starting to generate Genome files
May 22 15:06:58 ... starting to sort Suffix Array. This may take a long time...
May 22 15:08:18 ... sorting Suffix Array chunks and saving them to disk...
May 22 16:33:02 ... loading chunks from disk, packing SA...
May 22 16:34:41 ... finished generating suffix array
May 22 16:34:41 ... generating Suffix Array index
May 22 16:38:04 ... completed Suffix Array index
May 22 16:38:04 ..... processing annotations GTF
May 22 16:38:17 ..... inserting junctions into the genome indices
May 22 16:41:33 ... writing Genome to disk ...
May 22 16:41:48 ... writing Suffix Array to disk ...
May 22 16:42:01 ... writing SAindex to disk
May 22 16:42:02 ..... finished successfully
<Phase 2: Mapability Calculation>
May 22 16:42:02 ... mapping genome fragments back to genome...
May 22 18:09:16 ... sorting aligned genome fragments...
May 22 18:14:49 ... indexing aligned genome fragments...
May 22 18:15:47 ... filtering aligned genome fragments by chromosome/scaffold...
May 22 18:17:53 ... merging filtered genome fragments...
May 22 18:18:05 ... calculating regions for exclusion...
May 22 18:23:54 ... cleaning temporary files...
<Phase 3: IRFinder Reference Preparation>
May 22 18:23:58 ... building Ref 1...
May 22 18:24:25 ... building Ref 2...
May 22 18:24:27 ... building Ref 3...
May 22 18:24:27 ... building Ref 4...
May 22 18:25:02 ... building Ref 5...
May 22 18:25:42 ... building Ref 6...
May 22 18:25:43 ... building Ref 7...
May 22 18:25:44 ... building Ref 8...
May 22 18:25:45 ... building Ref 9...
May 22 18:25:45 ... building Ref 10b...
May 22 18:25:45 ... building Ref 11b...
May 22 18:25:45 ... building Ref 12b...`
and:
`$job_1722418.err
**[bam_sort_core] merging from 60 files and 30 in-memory blocks...**`
In adddition, there are no log files in the Mapability folder in v1.3.0. I still think the reference building is complete (the bed files in REF/genome/IRFinder are not empty), but the log from 1.2.6 was more informative.

**Comment by dg520** [comment]
Hi @mrashwing ,

Appreciate this throughout test! Yes, IRFinder 1.3.0 works now.

This difference between standard outputs of 1.2.6 and 1.3.0 (your log files) is because we've re-deisgned the mappability algorithm and also modified the message to make the three phases uniform.     
For the standard error messages (your `.err` files ), the 1.2.6 reports the running time for three steps during the mapability phase (Phase 2). In the 1.3.0, we've moved running time information to standard output and it is directly printed as local time in front of each step.    
For the log files in the `Mapability` folder in IRFinder 1.2.6, those are log files generated by STAR. It tells how genome fragments are mapped back against the genome. That information is not necessary for IRFinder but more for debug purpose, so we've removed them in 1.3.0 to only keep the files IRFinder requires.

Best,
Dadi
